sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -311,17 +324,26 @@ public OffsetRangeTracker restrictionTracker(
     if (restriction.getTo() < Long.MAX_VALUE) {
       return new OffsetRangeTracker(restriction);
     }
-    Map<String, Object> updatedConsumerConfig =
-        overrideBootstrapServersConfig(consumerConfig, kafkaSourceDescriptor);
-    KafkaLatestOffsetEstimator offsetPoller =
-        new KafkaLatestOffsetEstimator(
-            consumerFactoryFn.apply(
-                KafkaIOUtils.getOffsetConsumerConfig(
-                    "tracker-" + kafkaSourceDescriptor.getTopicPartition(),
-                    offsetConsumerConfig,
-                    updatedConsumerConfig)),
-            kafkaSourceDescriptor.getTopicPartition());
-    return new GrowableOffsetRangeTracker(restriction.getFrom(), offsetPoller);
+    final Map<TopicPartition, KafkaLatestOffsetEstimator> offsetEstimatorCacheInstance =
+        Preconditions.checkStateNotNull(this.offsetEstimatorCache);
+
+    TopicPartition topicPartition = kafkaSourceDescriptor.getTopicPartition();
+    KafkaLatestOffsetEstimator offsetEstimator = offsetEstimatorCacheInstance.get(topicPartition);
```

Comment:
The estimator should be cheap and use a consumer that is created outside of it, right?

Reply:
It is cheap -- but the reason why I didn't cache the Consumer is because `KafkaLatestOffsetEstimator` has the offsets memoized, which wouldn't do much unless you reuse the estimator.

Reply:
I agree -- but TopicPartition comes as an element in a PCollection, so changing the scope is more complicated and requires more deep thinking and testing.

There are some more fundamental changes that need to be done to make this happen -- but given the trouble that this can cause for a few clusters, this feels like a good patch to get in while we think about how to make SDF causing less overhead (I still see a bunch of short-lived consumers/connections for process continuation, which didn't happen in legacy).